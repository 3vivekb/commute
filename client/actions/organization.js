import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addOrganization = createAction('add organization')
export const createOrganization = (newOrganization) => {
  newOrganization.id = uuid.v4()
  newOrganization.sites = []
  newOrganization.groups = []
  return [
    addOrganization(newOrganization),
    push('/')
  ] // TODO save to server
}

// site stuff
const addSite = createAction('add site')
export const createSite = (newSite) => {
  newSite.id = uuid.v4()
  return [
    addSite(newSite),
    push(`/organizations/${newSite.organizationId}`)
  ] // TODO save to server
}
