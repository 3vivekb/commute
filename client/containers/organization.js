import {connect} from 'react-redux'

import Organization from '../components/organization'

function mapStateToProps (state, props) {
  const {organization} = state
  const {params} = props
  const currentOrganization = organization.organizationsById[params.organizationId]
  return {organization: currentOrganization}
}

export default connect(mapStateToProps)(Organization)
