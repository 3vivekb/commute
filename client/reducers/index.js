import {handleActions} from 'redux-actions'

import * as commuter from './commuter'
import * as multiSite from './multi-site'
import * as site from './site'
import * as user from './user'

export default {
  commuter: handleActions(commuter.reducers, commuter.initialState),
  multiSite: handleActions(multiSite.reducers, multiSite.initialState),
  site: handleActions(site.reducers, site.initialState),
  user: handleActions(user.reducers, user.initialState)
}
