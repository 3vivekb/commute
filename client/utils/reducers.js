import update from 'react-addons-update'

import {addToEntityMap, addEntitiesToEntityMap, deleteFromEntityMap, deleteManyFromEntityMap} from './entities'

function addChildEntityToParent (parentMap, newEntity, parentIdField, childPluralName) {
  return update(parentMap, {
    [newEntity[parentIdField]]: {
      [childPluralName]: {
        $push: [newEntity._id]
      }
    }
  })
}

/**
 * Make handlers for children creation and deletion events
 *
 * @param  {Object} cfg  An configuration object structured as follows:
 *   - childPluralName (String): Plural name of child (parent's child id array)
 *   - childSingularName (String): Singular name of child (for making reducer title)
 *   - parentIdField (String): Id field of parent in child entity
 * @return {Object}     Reducers
 */
export function makeChildrenHandlers (cfg) {
  const reducers = {}
  reducers[`add ${cfg.childSingularName}`] = function (state, action) {
    const newEntity = action.payload
    return addChildEntityToParent(
      state,
      newEntity,
      cfg.parentIdField,
      cfg.childPluralName
    )
  }

  reducers[`add many ${cfg.childPluralName}`] = function (state, action) {
    const newEntities = action.payload
    let updatedState = state
    newEntities.forEach(newEntity => {
      updatedState = addChildEntityToParent(
        updatedState,
        newEntity,
        cfg.parentIdField,
        cfg.childPluralName
      )
    })
    return updatedState
  }

  reducers[`delete ${cfg.childSingularName}`] = function (state, action) {
    const entity = action.payload
    const childIds = state[entity[cfg.parentIdField]][cfg.childPluralName]
    return update(state, {
      [entity[cfg.parentIdField]]: {
        [cfg.childPluralName]: {
          $set: childIds.filter((id) => id !== entity._id)
        }
      }
    })
  }

  return reducers
}

/**
 * Make generic reducers for an entity type
 *
 * @param  {Object} cfg  An object with name and handlers parameters structured as follows:
 *   - name: must have singular and plural keys with the corresponding strings
 *   - handlers: An array of strings of handlers.  Possible values: ['add', 'add many', 'delete', 'set', 'set many']
 * @return {Object}      The generated reducers
 */
export function makeGenericReducerHandlers (cfg) {
  const reducers = {
    'log out' (state, action) {
      return {}
    }
  }
  if (cfg.handlers.indexOf('add') !== -1) {
    reducers[`add ${cfg.name.singular}`] = function (state, action) {
      return addToEntityMap(state, action.payload)
    }
  }

  if (cfg.handlers.indexOf('add many') !== -1) {
    reducers[`add many ${cfg.name.plural}`] = function (state, action) {
      return addEntitiesToEntityMap(state, action.payload)
    }
  }

  if (cfg.handlers.indexOf('delete') !== -1) {
    reducers[`delete ${cfg.name.singular}`] = function (state, action) {
      return deleteFromEntityMap(state, action.payload._id)
    }
  }

  if (cfg.handlers.indexOf('delete many') !== -1) {
    reducers[`delete many ${cfg.name.plural}`] = function (state, action) {
      return deleteManyFromEntityMap(state, action.payload)
    }
  }

  if (cfg.handlers.indexOf('set') !== -1) {
    reducers[`set ${cfg.name.singular}`] = function (state, action) {
      return addToEntityMap(state, action.payload)
    }
  }

  if (cfg.handlers.indexOf('set many') !== -1) {
    reducers[`set ${cfg.name.plural}`] = function (state, action) {
      return addEntitiesToEntityMap(state, action.payload)
    }
  }

  return reducers
}
