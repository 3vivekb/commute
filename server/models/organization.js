const {Schema} = require('mongoose')

const Organization = module.exports = new Schema({
  createdBy: String,
  name: String
})

Organization.methods.update = function (properties) {
  Object
    .keys(properties)
    .forEach((key) => {
      this[key] = properties[key]
    })
  return this.save()
}
