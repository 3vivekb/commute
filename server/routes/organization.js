const models = require('../models')
const makeRestEndpoints = require('../utils/restEndpoints')

module.exports = function makeRoutes (app) {
  makeRestEndpoints(app,
    {
      childModels: [{
        foreignKey: 'organizationId',
        key: 'analyses',
        model: models.Analysis
      }, {
        foreignKey: 'organizationId',
        key: 'groups',
        model: models.Group
      }, {
        foreignKey: 'organizationId',
        key: 'sites',
        model: models.Site
      }],
      commands: {
        'Collection GET': {},
        'Collection POST': {},
        'GET': {},
        'DELETE': {},
        'PUT': {}
      },
      model: models.Organization,
      name: 'organization'
    }
  )
}
