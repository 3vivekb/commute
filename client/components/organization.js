import React, {Component, PropTypes} from 'react'
import {Button, ButtonGroup, Col, Grid, Row} from 'react-bootstrap'
import {BootstrapTable, TableHeaderColumn} from 'react-bootstrap-table'
import {Link} from 'react-router'

import Icon from './icon'
import {messages} from '../utils/env'
import {arrayCountRenderer} from '../utils/table'
import {actUponConfirmation} from '../utils/ui'

export default class Organization extends Component {
  static propTypes = {
    // dispatch
    deleteAnalysis: PropTypes.func,
    deleteGroup: PropTypes.func,
    deleteOrganization: PropTypes.func,
    deleteSite: PropTypes.func,

    // props
    analysis: PropTypes.object.isRequired,
    analyses: PropTypes.array.isRequired,
    group: PropTypes.object.isRequired,
    groups: PropTypes.array.isRequired,
    organization: PropTypes.object.isRequired,
    site: PropTypes.object.isRequired,
    sites: PropTypes.array.isRequired
  }

  _analysisGroupNameRenderer = (cell, row) => {
    const group = this.props.group[row.groupId]
    return <Link to={`/group/${group.id}`}>{group.name}</Link>
  }

  _analysisNameRenderer = (cell, row) => {
    return <Link to={`/analysis/${row.id}`}>{row.name}</Link>
  }

  _analysisSiteNameRenderer = (cell, row) => {
    const site = this.props.site[row.siteId]
    return <Link to={`/group/${site.id}`}>{site.name}</Link>
  }

  _analysisToolsRenderer = (cell, row) => {
    const doDelete = () => {
      this.props.deleteAnalysis({ id: row.id, organizationId: this.props.organization.id })
    }
    const onClick = () => actUponConfirmation(messages.analysis.deleteConfirmation, doDelete)
    return <Button bsStyle='danger' onClick={onClick}>Delete</Button>
  }

  _handleDelete = () => {
    const doDelete = () => this.props.deleteOrganization(this.props.organization)
    actUponConfirmation(messages.organization.deleteConfirmation, doDelete)
  }

  _groupNameRenderer = (cell, row) => {
    return <Link to={`/group/${row.id}/`}>{row.name}</Link>
  }

  _groupToolsRenderer = (cell, row) => {
    const doDelete = () => this.props.deleteGroup({ id: row.id, organizationId: this.props.organization.id })
    const onClick = () => actUponConfirmation(messages.group.deleteConfirmation, doDelete)
    return <Button bsStyle='danger' onClick={onClick}>Delete</Button>
  }

  _siteNameRenderer = (cell, row) => {
    return <Link to={`/site/${row.id}/edit`}>{row.name}</Link>
  }

  _siteToolsRenderer = (cell, row) => {
    const organizationId = this.props.organization.id
    const doDelete = () => this.props.deleteSite({ id: row.id, organizationId })
    const onClick = () => actUponConfirmation(messages.site.deleteConfirmation, doDelete)
    return <div>
      <Button bsStyle='warning'>
        <Link to={`/site/${row.id}/edit`}>Edit</Link>
      </Button>
      <Button bsStyle='danger' onClick={onClick}>Delete</Button>
    </div>
  }

  render () {
    const {analyses, groups, sites} = this.props
    const {id: organizationId, name} = this.props.organization

    return (
      <Grid>
        <Row>
          <Col xs={12}>
            <h2><Icon type='shield' />{name}</h2>
            <ButtonGroup>
              <Button bsStyle='warning'>
                <Link to={`/organization/${organizationId}/edit`}>Edit</Link>
              </Button>
              <Button bsStyle='danger' onClick={this._handleDelete}>Delete</Button>
            </ButtonGroup>
            <p>Below are this organization's sites, commuter groups and analyses.</p>
            <h3>Sites
              <Button className='pull-right'>
                <Link to={`/organization/${organizationId}/site/create`}>Create a new site <Icon type='building' /></Link>
              </Button>
            </h3>
            <p>A site is a location of a building or new address that you want to use as the centerpoint of your commutes.</p>
            <BootstrapTable data={sites}>
              <TableHeaderColumn dataField='id' isKey hidden />
              <TableHeaderColumn dataFormat={this._siteNameRenderer}>Name</TableHeaderColumn>
              <TableHeaderColumn dataField='address'>Address</TableHeaderColumn>
              <TableHeaderColumn dataFormat={this._siteToolsRenderer}>Tools</TableHeaderColumn>
            </BootstrapTable>
            <h3>Commuter Groups
              <Button className='pull-right'>
                <Link to={`/organization/${organizationId}/group/create`}>Create a new group <Icon type='users' /></Link>
              </Button>
            </h3>
            <p>A commuter group is a list of commuters that can commute to a particular site.</p>
            <BootstrapTable data={groups}>
              <TableHeaderColumn dataField='id' isKey hidden />
              <TableHeaderColumn dataFormat={this._groupNameRenderer}>Name</TableHeaderColumn>
              <TableHeaderColumn dataField='commuters' dataFormat={arrayCountRenderer}>Commuters</TableHeaderColumn>
              <TableHeaderColumn dataFormat={this._groupToolsRenderer}>Tools</TableHeaderColumn>
            </BootstrapTable>
            <h3>Analyses
              <Button className='pull-right'>
                <Link to={`/organization/${organizationId}/analysis/create`}>Create a new analysis <Icon type='bar-chart' /></Link>
              </Button>
            </h3>
            <p>An analysis calculates commuting statistics for a pairing of a commuter group and site.</p>
            <BootstrapTable data={analyses}>
              <TableHeaderColumn dataField='id' isKey hidden />
              <TableHeaderColumn dataFormat={this._analysisNameRenderer}>Name</TableHeaderColumn>
              <TableHeaderColumn dataFormat={this._analysisSiteNameRenderer}>Site</TableHeaderColumn>
              <TableHeaderColumn dataFormat={this._analysisGroupNameRenderer}>Group</TableHeaderColumn>
              <TableHeaderColumn dataFormat={this._analysisToolsRenderer}>Tools</TableHeaderColumn>
            </BootstrapTable>
          </Col>
        </Row>
      </Grid>
    )
  }
}
