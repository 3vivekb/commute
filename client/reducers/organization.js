import {makeChildrenHandlers, makeGenericReducerHandlers} from '../utils/reducers'

export const reducers = makeGenericReducerHandlers({
  handlers: ['add', 'delete', 'set', 'set all'],
  name: {
    singular: 'organization',
    plural: 'organizations'
  }
})

Object.assign(reducers, makeChildrenHandlers({
  childPluralName: 'analyses',
  childSingularName: 'analysis',
  parentIdField: 'organizationId'
}))

Object.assign(reducers, makeChildrenHandlers({
  childPluralName: 'groups',
  childSingularName: 'group',
  parentIdField: 'organizationId'
}))

Object.assign(reducers, makeChildrenHandlers({
  childPluralName: 'sites',
  childSingularName: 'site',
  parentIdField: 'organizationId'
}))

export const initialState = {}
