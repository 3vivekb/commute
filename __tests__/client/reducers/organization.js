/* globals describe */

import {blankOrganization, mockStores, organizationWithAnAnalysis} from '../../test-utils/mock-data'
import {makeChildrenHandlerTestCases, makeGenericReducerTestCases} from '../../test-utils/reducers'

import * as organization from '../../../client/reducers/organization'

describe('client > reducers > organization', () => {
  makeChildrenHandlerTestCases({
    add: {
      affectedParentId: 'organization-1',
      initialState: mockStores.withBlankOrganization.organization,
      payload: { _id: 'analysis-new', organizationId: 'organization-1' }
    },
    childPluralName: 'analyses',
    childSingularName: 'analysis',
    delete: {
      affectedParentId: 'organization-2',
      initialState: mockStores.withAnalysisRun.organization,
      payload: { _id: 'analysis-2', organizationId: 'organization-2' }
    },
    initialState: organization.initialState,
    reducers: organization.reducers
  })

  makeChildrenHandlerTestCases({
    add: {
      affectedParentId: 'organization-1',
      initialState: mockStores.withBlankOrganization.organization,
      payload: { _id: 'group-new', organizationId: 'organization-1' }
    },
    childPluralName: 'groups',
    childSingularName: 'group',
    delete: {
      affectedParentId: 'organization-2',
      initialState: mockStores.withAnalysisRun.organization,
      payload: { _id: 'group-2', organizationId: 'organization-2' }
    },
    initialState: organization.initialState,
    reducers: organization.reducers
  })

  makeChildrenHandlerTestCases({
    add: {
      affectedParentId: 'organization-1',
      initialState: mockStores.withBlankOrganization.organization,
      payload: { _id: 'site-new', organizationId: 'organization-1' }
    },
    childPluralName: 'sites',
    childSingularName: 'site',
    delete: {
      affectedParentId: 'organization-2',
      initialState: mockStores.withAnalysisRun.organization,
      payload: { _id: 'site-2', organizationId: 'organization-2' }
    },
    initialState: organization.initialState,
    reducers: organization.reducers
  })

  makeGenericReducerTestCases({
    handlers: {
      add: {
        initialState: organization.initialState,
        payload: blankOrganization
      },
      delete: {
        initialState: mockStores.withBlankOrganization.organization,
        payload: 'organization-1'
      },
      set: {
        initialState: mockStores.withBlankOrganization.organization,
        payload: organizationWithAnAnalysis
      },
      'set many': {
        initialState: organization.initialState,
        payload: [blankOrganization, organizationWithAnAnalysis]
      }
    },
    initialState: organization.initialState,
    name: {
      plural: 'organizations',
      singular: 'organization'
    },
    reducers: organization.reducers
  })
})
