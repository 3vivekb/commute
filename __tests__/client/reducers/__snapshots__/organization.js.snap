exports[`client > reducers > organization should handle add organization 1`] = `
Object {
  "organizations": Array [
    Object {
      "id": 1,
    },
  ],
  "organizationsById": Object {
    "1": Object {
      "id": 1,
    },
  },
}
`;

exports[`client > reducers > organization should handle add site 1`] = `
Object {
  "organizations": Array [
    Object {
      "analyses": Array [],
      "analysesById": Object {},
      "contact": "someone",
      "email": "abc@def.ghi",
      "groups": Array [],
      "groupsById": Object {},
      "id": "1",
      "logo_url": "https://placekitten.com/80/80",
      "main_url": "https://placekitten.com/",
      "name": "Mock Organization",
      "sites": Array [
        Object {
          "address": "123 ABC St",
          "id": "1",
          "name": "Acme Corp",
          "organizationId": "1",
        },
      ],
      "sitesById": Object {
        "1": Object {
          "address": "123 ABC St",
          "id": "1",
          "name": "Acme Corp",
          "organizationId": "1",
        },
      },
    },
  ],
  "organizationsById": Object {
    "1": Object {
      "analyses": Array [],
      "analysesById": Object {},
      "contact": "someone",
      "email": "abc@def.ghi",
      "groups": Array [],
      "groupsById": Object {},
      "id": "1",
      "logo_url": "https://placekitten.com/80/80",
      "main_url": "https://placekitten.com/",
      "name": "Mock Organization",
      "sites": Array [
        Object {
          "address": "123 ABC St",
          "id": "1",
          "name": "Acme Corp",
          "organizationId": "1",
        },
      ],
      "sitesById": Object {
        "1": Object {
          "address": "123 ABC St",
          "id": "1",
          "name": "Acme Corp",
          "organizationId": "1",
        },
      },
    },
  },
}
`;

exports[`client > reducers > organization should handle default state 1`] = `
Object {
  "organizations": Array [],
  "organizationsById": Object {},
}
`;

exports[`client > reducers > organization should handle set organizations 1`] = `
Object {
  "organizations": Array [
    Array [
      Object {
        "id": 1,
      },
    ],
  ],
  "organizationsById": Object {
    "1": Object {
      "id": 1,
    },
  },
}
`;
